packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,793 @@
+    const blockInput = new BlockInputColumns(init, state, props.custodyConfig);
+    if (hasAllData) {
+      blockInput.dataPromise.resolve([]);
+    }
+    return blockInput;
+  }
+
+  getLogMeta(): LogMetaColumns {
+    return {
+      blockRoot: prettyBytes(this.blockRootHex),
+      slot: this.slot,
+      expectedColumns:
+        this.state.hasBlock && this.state.block.message.body.blobKzgCommitments.length === 0
+          ? 0
+          : this.custodyConfig.sampledColumns.length,
+      receivedColumns: this.getSampledColumns().length,
+    };
+  }
+
+  addBlock(props: AddBlock<ForkColumnsDA>): void {
+    if (this.state.hasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INVALID_CONSTRUCTION,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot addBlock to BlockInputColumns after it already has a block"
+      );
+    }
+
+    if (props.blockRootHex !== this.blockRootHex) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISMATCHED_ROOT_HEX,
+          blockInputRoot: this.blockRootHex,
+          mismatchedRoot: props.blockRootHex,
+          source: props.source.source,
+          peerId: `${props.source.peerIdStr}`,
+        },
+        "addBlock blockRootHex does not match BlockInput.blockRootHex"
+      );
+    }
+
+    for (const {columnSidecar} of this.columnsCache.values()) {
+      if (!blockAndColumnArePaired(props.block, columnSidecar)) {
+        this.columnsCache.delete(columnSidecar.index);
+        // this.logger?.error(`Removing columnIndex=${columnSidecar.index} from BlockInput`, {}, err);
+      }
+    }
+
+    const hasAllData = props.block.message.body.blobKzgCommitments.length === 0 || this.state.hasAllData;
+
+    this.state = {
+      ...this.state,
+      hasBlock: true,
+      hasAllData,
+      block: props.block,
+      source: props.source,
+      timeCompleteSec: hasAllData ? props.source.seenTimestampSec : undefined,
+    } as BlockInputColumnsState;
+
+    this.blockPromise.resolve(props.block);
+  }
+
+  addColumn({blockRootHex, columnSidecar, source, seenTimestampSec, peerIdStr}: AddColumn): void {
+    if (blockRootHex !== this.blockRootHex) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISMATCHED_ROOT_HEX,
+          blockInputRoot: this.blockRootHex,
+          mismatchedRoot: blockRootHex,
+          source: source,
+          peerId: `${peerIdStr}`,
+        },
+        "Column BeaconBlockHeader blockRootHex does not match BlockInput.blockRootHex"
+      );
+    }
+
+    if (this.state.hasBlock) {
+      assertBlockAndColumnArePaired(this.blockRootHex, this.state.block, columnSidecar);
+    }
+
+    this.columnsCache.set(columnSidecar.index, {columnSidecar, source, seenTimestampSec, peerIdStr});
+
+    const sampledColumns = this.getSampledColumns();
+    const hasAllData = this.state.hasAllData || sampledColumns.length === this.custodyConfig.sampledColumns.length;
```

Comment:
With dynamic custody in mind i think we should run this check and not rely on the state.hasAllData

```suggestion
    const hasAllData = sampledColumns.length === this.custodyConfig.sampledColumns.length;
```

Reply:
We shouldn't be dealing with an updating object here. We should make a copy of the sampled/custodied columns on block input creation (or pass in copies directly) so updates don't affect the single block input.

Sampling updates can be handled per-slot when we're synced at the tip. (A custody update can take effect the next slot).
Sampling updates can be handled per-batch when we're range syncing. (A custody update can take effect the next batch).
And in either case, the update may trigger a data backfill.